src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -125,114 +126,248 @@ EmbeddedAssemblies::TypeMappingInfo_compare_key (const void *a, const void *b)
 	return strcmp (reinterpret_cast <const char*> (a), reinterpret_cast <const char*> (b));
 }
 
-inline const char*
-EmbeddedAssemblies::find_entry_in_type_map (const char *name, uint8_t map[], TypeMapHeader& header)
+template<typename Key, typename Entry, int (*compare)(const Key*, const Entry*), bool use_extra_size>
+const Entry*
+EmbeddedAssemblies::binary_search (const Key *key, const Entry *base, size_t nmemb, [[maybe_unused]] size_t extra_size)
 {
-	const char *e = reinterpret_cast<const char*> (bsearch (name, map, header.entry_count, header.entry_length, TypeMappingInfo_compare_key ));
-	if (e == nullptr)
-		return nullptr;
-	return e + header.value_offset;
+	static_assert (compare != nullptr, "compare is a required template parameter");
+
+	constexpr size_t size = sizeof(Entry);
+	while (nmemb > 0) {
+		const Entry *ret;
+		if constexpr (use_extra_size) {
+			ret = reinterpret_cast<const Entry*>(reinterpret_cast<const uint8_t*>(base) + (size + extra_size) * (nmemb / 2));
+		} else {
+			ret = base + (nmemb / 2);
+		}
+
+		int result = compare (key, ret);
+		if (result < 0) {
+			nmemb /= 2;
+		} else if (result > 0) {
+			if constexpr (use_extra_size) {
+				base = reinterpret_cast<const Entry*>(reinterpret_cast<const uint8_t*>(ret) + size + extra_size);
+			} else {
+				base = ret + 1;
+			}
+			nmemb -= nmemb / 2 + 1;
+		} else {
+			return ret;
+		}
+	}
+
+	return nullptr;
 }
 
-const char*
-EmbeddedAssemblies::typemap_java_to_managed (const char *java)
+MonoReflectionType*
+EmbeddedAssemblies::typemap_java_to_managed (MonoString *java_type)
 {
-#if defined (DEBUG) || !defined (ANDROID)
-	for (TypeMappingInfo *info = java_to_managed_maps; info != nullptr; info = info->next) {
-		/* log_warn (LOG_DEFAULT, "# jonp: checking file: %s!%s for type '%s'", info->source_apk, info->source_entry, java); */
-		const char *e = reinterpret_cast<const char*> (bsearch (java, info->mapping, static_cast<size_t>(info->num_entries), static_cast<size_t>(info->entry_length), TypeMappingInfo_compare_key));
-		if (e == nullptr)
-			continue;
-		return e + info->value_offset;
+	timing_period total_time;
+	if (XA_UNLIKELY (utils.should_log (LOG_TIMING))) {
+		timing = new Timing ();
+		total_time.mark_start ();
 	}
+
+	simple_pointer_guard<char[], false> java_type_name (mono_string_to_utf8 (java_type));
```

Comment:
This "begs" an interesting question: since we control the file format, why use UTF-8 strings for the Java type names?  Why not use UTF-16 strings?  This way we wouldn't need to allocate a UTF-8 string for binary search.

Yes, it would ~double the struct sizes vs. what they currently are -- especially since it's *highly* unlikely Java types will ever be outside of ASCII -- but it would means fewer memory allocations at runtime.

Reply:
Yeah, it's an idea... The entire app would use more resident RAM and would be bigger on disk, but the code should run faster...